apigen/hook-docs.php

```diff
@@ -0,0 +1,377 @@
+	 *
+	 * @param  string  $pattern The pattern to match
+	 * @param  bool    $flags   Sort flags
+	 * @param  string  $path    Path to file
+	 *
+	 * @return array
+	 *
+	 * @since NEXT
+	 */
+	private static function get_files( $pattern, $flags = 0, $path = '' ) {
+
+		if ( ! $path && ( $dir = dirname( $pattern ) ) != '.' ) {
+
+			if ( '\\' == $dir || '/' == $dir ) {
+
+				$dir = '';
+
+			}
+
+			return self::get_files( basename( $pattern ), $flags, $dir . '/' );
+
+		}
+
+		$paths = glob( $path . '*', GLOB_ONLYDIR | GLOB_NOSORT );
+		$files = glob( $path . $pattern, $flags );
+
+		if ( is_array( $paths ) ) {
+
+			foreach ( $paths as $p ) {
+
+				$found_files    = array();
+				$retrieved_files = (array) self::get_files( $pattern, $flags, $p . '/' );
+
+				foreach ( $retrieved_files as $file ) {
+
+					if ( in_array( $file, self::$found_files ) ) {
+
+						continue;
+
+					}
+
+					$found_files[] = $file;
+
+				}
+
+				self::$found_files = array_merge( self::$found_files, $found_files );
+
+				if ( is_array( $files ) && is_array( $found_files ) ) {
+
+					$files = array_merge( $files, $found_files );
+
+				}
+
+			}
+
+		}
+
+		return $files;
+
+	}
+
+	/**
+	 * Get specific hook link
+	 *
+	 * @param  string $hook     Name of hook
+	 * @param  array  $details  Hook data array
+	 *
+	 * @return mixed
+	 *
+	 * @since NEXT
+	 */
+	private static function get_hook_link( $hook, $details = array() ) {
+
+		if ( ! empty( $details['class'] ) ) {
+
+			$link = "source-function-{$details['class']}.html#{$details['line']}";
+
+		} elseif ( ! empty( $details['function'] ) ) {
+
+			$link = "source-function-{$details['function']}.html#{$details['line']}";
+
+		} else {
+
+			$link = "https://github.com/godaddy/wp-primer-theme/search?utf8=%E2%9C%93&q={$hook}";
+
+		}
+
+		return '<a href="' . $link . '">' . $hook . '</a>';
+
+	}
+
+	/**
+	 * Process files, pull hooks & render table
+	 *
+	 * @return mixed
+	 */
+	public static function process_hooks() {
+
+		$template_files   = self::get_files( '*.php', GLOB_MARK, '../templates/parts/' );
+
+		$other_files     = array(
+			'../inc/hooks.php',
+			'../inc/helpers.php',
+			'../inc/template-tags.php',
+		);
+
+		self::$files_to_scan = array(
+			'Template Hooks'  => $template_files,
+			'Other Hooks'     => $other_files,
+		);
+
+		$scanned = array();
+
+		ob_start();
+
+		echo '<div id="content">';
+		echo '<h1>Action and Filter Hook Reference</h1>';
+
+		foreach ( self::$files_to_scan as $heading => $files ) {
+
+			self::$custom_hooks_found = array();
+
+			foreach ( $files as $f ) {
+
+				self::$current_file = basename( $f );
+				$tokens             = token_get_all( file_get_contents( $f ) );
+				$token_type         = false;
+				$current_class      = '';
+				$current_function   = '';
+
+				if ( in_array( self::$current_file, $scanned ) ) {
+
+					continue;
+
+				}
+
+				$scanned[] = self::$current_file;
+
+				foreach ( $tokens as $index => $token ) {
```

Comment:
@EvanHerman Where was this taken from? The amount of indentation is pretty high here.